### PR TITLE
Editing the title in the front matter

### PR DESCRIPTION
**Pull request by edditler**
- Rename `update_item` -> `update_item_md`
- Add `api.update_dataobj_frontmatter`
- Add `data.update_item_frontmatter`. This routine only changes the elements that were supplied to it. This way it can be also used to change date, type and tags.
- Remove `title` from `show.html` because we already have `dataobj['title']`
- Adapt the CSS of the frontmatter a bit so that everything is aligned on the left.

### CODE REVIEW

**Comment by edditler** [comment]
I'm not sure how to run the test locally, but this should work

**Comment by Uzay-G** [comment]
Don't worry I'll figure out the problems with the tests and all :+1:. I'll put some comments and suggestions too.

File: archivy/api.py [hunk 1]
```diff
@@ -124,13 +124,34 @@ def update_dataobj(dataobj_id):
...
 
 
+@api_bp.route("/dataobjs/frontmatter/<int:dataobj_id>", methods=["PUT"])
+def update_dataobj_frontmatter(dataobj_id):
+    """
+    Updates frontmatter of object of given id.
+
+    Paramter in JSON body:
+
+    - **content**: markdown text of new dataobj.
```
Comment (Uzay-G) [inline archivy/api.py:141]: We need to change this from content -> title. Do you think we should also be able to edit the tags this way?

**Comment by Uzay-G** [comment]
I really like the styling with the tags, it's very nice :ok_hand:. I think maybe it should be more clear that the form has submitted when you enter or click the button, but I'm not sure how we should go about signaling that? Maybe the check mark could change to green or something.

File: archivy/routes.py [hunk 2]
```diff
@@ -121,15 +121,18 @@ def show_dataobj(dataobj_id):
                 if hit["id"] != dataobj_id:
                     backlinks.append({"title": hit["title"], "id": hit["id"]})
 
+    post_title_form = forms.TitleForm()
+    post_title_form.title.data = dataobj["title"]
+
     return render_template(
         "dataobjs/show.html",
-        title=dataobj["title"],
```
Comment (Uzay-G) [inline archivy/routes.py:126]: The way the template works is the title attribute fills the `<title>` tag of the returned html, and this is true for all routes, so we need it. Don't worry, I'll add it back :)

**Comment by Uzay-G** [comment]
I find the small black border between the buttons and tags a bit jarring, what do you think?

File: archivy/templates/dataobjs/show.html [hunk 3]
```diff
@@ -17,36 +17,55 @@
...
 	{% endif %}
-    <h2 id="post-title">{{ dataobj['title'] }}</h2>
-    {{ ", ".join(dataobj['tags']) }}
-    <br>
+  <div class="post-header">
+    <!-- Form for editing the frontmatter -->
+    {% if not view_only %}
+    <form id="post-title-form">
+      {{ post_title_form.hidden_tag() }}
+        {{ post_title_form.title(**{"aria-describedby":"username-input-validation"}) }}
```
Comment (Uzay-G) [inline archivy/templates/dataobjs/show.html:25]: Could you explain what the `aria-describedby` here does?

File: archivy/api.py (hunk 1 above)
Comment (edditler) [reply archivy/api.py:141]: Thing with tags is that a `tags.split(",")` is not robust on its own and the user needs to know that commas are separators (as opposed to spaces for example). Maybe when work on the tag systems starts there might be a nicer way (autosuggest existing tags and stuff like that). The API supports changing the tags already.

File: archivy/routes.py (hunk 2 above)
Comment (edditler) [reply archivy/routes.py:126]: sure

File: archivy/templates/dataobjs/show.html (hunk 3 above)
Comment (edditler) [reply archivy/templates/dataobjs/show.html:25]: Ah sorry I was copying and pasting.

**Comment by edditler** [comment]
I have the symbol turn green now after saving. This is a general thing with all async requests. The `flash()` are only visible on reload. Maybe there should be some way of flashing JS notifications?

**Comment by edditler** [comment]
> I find the small black border between the buttons and tags a bit jarring, what do you think?

You mean the border below the buttons? I put that in to separate the post-header and post-content. I don't have a strong opinion on the design, so feel free to change it.

File: archivy/data.py [hunk 4]
```diff
@@ -193,6 +193,40 @@ def update_item(dataobj_id, new_content):
...
+
+    filename = get_by_id(dataobj_id)
+    dataobj = frontmatter.load(filename)
+    for key in list(new_frontmatter):
+        dataobj[key] = new_frontmatter[key]
+    md = frontmatter.dumps(dataobj)
+    with open(filename, "w", encoding="utf-8") as f:
+        f.write(md)
+
+    # TODO: What does this do?
```
Comment (Uzay-G) [inline archivy/data.py:221]: This basically sends out a hook where the user can call code on the newly edited item (https://archivy.github.io/reference/hooks/).

I'll refactor this in another PR though

**Comment by Uzay-G** [comment]
Perfect!

**Pull request merged**